src/Compilers/CSharp/Portable/Binder/PatternVariableFinder.cs

```diff
@@ -13,51 +13,42 @@ namespace Microsoft.CodeAnalysis.CSharp
     class PatternVariableFinder : CSharpSyntaxWalker
     {
         ArrayBuilder<DeclarationPatternSyntax> declarationPatterns;
-        ArrayBuilder<ExpressionSyntax> expressionsToVisit = ArrayBuilder<ExpressionSyntax>.GetInstance();
+        ArrayBuilder<CSharpSyntaxNode> nodesToVisit = ArrayBuilder<CSharpSyntaxNode>.GetInstance();
         internal static ArrayBuilder<DeclarationPatternSyntax> FindPatternVariables(
-            ExpressionSyntax expression = null,
-            ImmutableArray<ExpressionSyntax> expressions = default(ImmutableArray<ExpressionSyntax>),
-            ImmutableArray<PatternSyntax> patterns = default(ImmutableArray<PatternSyntax>))
+            CSharpSyntaxNode node = null,
+            ImmutableArray<CSharpSyntaxNode> nodes = default(ImmutableArray<CSharpSyntaxNode>))
         {
             var finder = s_poolInstance.Allocate();
             finder.declarationPatterns = ArrayBuilder<DeclarationPatternSyntax>.GetInstance();
 
-            // push expressions to be visited onto a stack
-            var expressionsToVisit = finder.expressionsToVisit;
-            if (expression != null) expressionsToVisit.Add(expression);
-            if (!expressions.IsDefaultOrEmpty)
+            // push nodes to be visited onto a stack
+            var nodesToVisit = finder.nodesToVisit;
+            if (node != null) nodesToVisit.Add(node);
+            if (!nodes.IsDefaultOrEmpty)
             {
-                foreach (var subExpression in expressions)
+                foreach (var subExpression in nodes)
                 {
-                    if (subExpression != null) expressionsToVisit.Add(subExpression);
+                    if (subExpression != null) nodesToVisit.Add(subExpression);
                 }
             }
 
-            expressionsToVisit.ReverseContents();
+            nodesToVisit.ReverseContents();
```

Comment:
Good observation. This code, however, is essentially obsolete, there is a pending PR (https://github.com/dotnet/roslyn/pull/9396) that eliminates this ArrayBuilder. The code is doing what it is supposed to do and perf is not a priority for it for the preview.